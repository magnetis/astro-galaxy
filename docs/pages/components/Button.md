# Button

Buttons define what action will occur when the user interacts with features and components of the platform. Buttons are used to initialize and dictate the actions on the interface, either in the background or foreground of an experience.

## primary buttons

Primary buttons are medium sized and have a colored background. They aren't always the same color, though, so there isn't a default "primary" look or CSS class - primary here is more of a concept.

Please note that color and size are built through CSS classes applied to the `<button>` element. You'll always need to specify color and size to get a complete button.

Use these for the main action in a section or screen. You should display only one primary button at a time.

Mouseover and click to view :hover and :active states.

```.jsx
<Button>Button</Button>
```
