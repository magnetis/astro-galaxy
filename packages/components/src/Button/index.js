import styled from 'styled-components';
import { color, space, border, layout, fontSize, buttonStyle, typography } from 'styled-system';
import { font, buttonSizes } from '@magnetis/astro-galaxy-core';

const hoverStyles = props => ({
  'primary.uranus': `
    background-color: ${props.theme.colors.uranus300};
    border-color: ${props.theme.colors.uranus300};
  `,
  'primary.earth': `
    background-color: ${props.theme.colors.earth200};
    border-color: ${props.theme.colors.earth200};
  `,
  'primary.venus': `
    background-color: ${props.theme.colors.venus200};
    border-color: ${props.theme.colors.venus200};
  `,
  'primary.mars': `
    background-color: ${props.theme.colors.mars300};
    border-color: ${props.theme.colors.mars300};
  `,
});

const activeStyles = props => ({
  'primary.uranus': `
    background-color: ${props.theme.colors.uranus600};
    border-color: ${props.theme.colors.uranus600};
  `,
  'primary.earth': `
    background-color: ${props.theme.colors.earth500};
    border-color: ${props.theme.colors.earth500};
  `,
  'primary.venus': `
    background-color: ${props.theme.colors.venus500};
    border-color: ${props.theme.colors.venus500};
  `,
  'primary.mars': `
    background-color: ${props.theme.colors.mars600};
    border-color: ${props.theme.colors.mars600};
  `,
});

const Button = styled.button`
  cursor: pointer;
  white-space: nowrap;
  text-align: center;
  transition: all 0.3s ease;
  overflow: hidden;
  border-width: 2px;
  border-style: solid;

  :disabled,
  .disabled {
    cursor: not-allowed;
    background-color: ${props => props.theme.colors.moon200};
    border-color: ${props => props.theme.colors.moon200};
    color: ${props => props.theme.colors.space100};
  }

  :hover:not(:disabled):not(.disabled),
  :focus:not(:disabled):not(.disabled) {
    ${props => hoverStyles(props)[props.variant]}
  }

  :active:not(:disabled):not(.disabled) {
    ${props => activeStyles(props)[props.variant]}
  }

  :focus {
    outline: none;
  }

  ${font};
  ${color};
  ${space};
  ${border};
  ${layout};
  ${fontSize};
  ${typography};
  ${buttonStyle};
  ${buttonSizes};
`;

Button.defaultProps = {
  display: 'inline-block',
  variant: 'primary.uranus',
  buttonSize: 'medium',
  font: 'primary',
};

Button.propTypes = {
  ...color.propTypes,
  ...space.propTypes,
  ...border.propTypes,
  ...layout.propTypes,
  ...fontSize.propTypes,
  ...buttonStyle.propTypes,
  ...typography.propTypes,
};

export default Button;
