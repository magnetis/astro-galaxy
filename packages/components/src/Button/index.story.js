import React from 'react';
import { storiesOf } from "@storybook/react";
import Button from './index.js';

storiesOf('Buttons', module)
  .add('button default', () => <Button>uranus</Button>)
  .add('primary buttons variants', () => (
    <React.Fragment>
      <Button variant='primary.uranus'>uranus</Button>
      <Button variant='primary.earth'>earth</Button>
      <Button variant='primary.venus'>venus</Button>
      <Button variant='primary.mars'>mars</Button>
    </React.Fragment>
  ));

storiesOf('Button small', module)
  .add('button uranus small', () => <Button buttonSize='small'>uranus</Button>);

storiesOf('Button large', module)
  .add('button uranus large', () => <Button buttonSize='large'>uranus</Button>);
