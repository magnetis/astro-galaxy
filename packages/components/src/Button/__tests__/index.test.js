import React from 'react';
import { theme } from '@magnetis/astro-galaxy-core';
import Button from '../index';

describe('Button component', () => {
  it('should render a button', () => {
    const button = rendererCreateWithTheme(<Button>Hello</Button>).toJSON();

    expect(button).toMatchSnapshot();
  });

  it('should change font-size and padding when button size is small', () => {
    const button = rendererCreateWithTheme(
      <Button buttonSize="small">Click here!</Button>
    ).toJSON();

    expect(button).toMatchSnapshot();
    expect(button).toHaveStyleRule('font-size', '12px');
    expect(button).toHaveStyleRule('padding', '5px 28px');
  });

  it('should change background-color and border-color when variant is primary.mars', () => {
    const button = rendererCreateWithTheme(
      <Button variant="primary.mars">Click here!</Button>
    ).toJSON();

    expect(button).toMatchSnapshot();
    expect(button).toHaveStyleRule('background-color', theme.colors.mars500);
    expect(button).toHaveStyleRule('border-color', theme.colors.mars500);
  });
});
