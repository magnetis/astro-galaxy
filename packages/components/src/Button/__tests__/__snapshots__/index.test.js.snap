// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should change background-color and border-color when variant is primary.mars 1`] = `
.c0 {
  cursor: pointer;
  white-space: nowrap;
  text-align: center;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  overflow: hidden;
  border-width: 2px;
  border-style: solid;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: normal normal 600 normal 16px/1.5 Poppins,sans-serif;
  display: inline-block;
  background-color: #d45459;
  color: #ffffff;
  border-color: #d45459;
  padding: 8px 38px;
  font-size: 16px;
  line-height: 28px;
  border-radius: 24px;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e9eef2;
  border-color: #e9eef2;
  color: #ffffff;
}

.c0:hover:not(:disabled),
.c0:focus:not(:disabled) {
  background-color: #ea9297;
  border-color: #ea9297;
}

.c0:active:not(:disabled) {
  background-color: #ac3c40;
  border-color: #ac3c40;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  display="inline-block"
>
  Click here!
</button>
`;

exports[`Button component should change font-size and padding when button size is small 1`] = `
.c0 {
  cursor: pointer;
  white-space: nowrap;
  text-align: center;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  overflow: hidden;
  border-width: 2px;
  border-style: solid;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: normal normal 600 normal 16px/1.5 Poppins,sans-serif;
  display: inline-block;
  background-color: #159ce4;
  color: #ffffff;
  border-color: #159ce4;
  padding: 5px 28px;
  font-size: 12px;
  line-height: 18px;
  border-radius: 16px;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e9eef2;
  border-color: #e9eef2;
  color: #ffffff;
}

.c0:hover:not(:disabled),
.c0:focus:not(:disabled) {
  background-color: #67cff3;
  border-color: #67cff3;
}

.c0:active:not(:disabled) {
  background-color: #147ec1;
  border-color: #147ec1;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  display="inline-block"
>
  Click here!
</button>
`;

exports[`Button component should render a button 1`] = `
.c0 {
  cursor: pointer;
  white-space: nowrap;
  text-align: center;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  overflow: hidden;
  border-width: 2px;
  border-style: solid;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: normal normal 600 normal 16px/1.5 Poppins,sans-serif;
  display: inline-block;
  background-color: #159ce4;
  color: #ffffff;
  border-color: #159ce4;
  padding: 8px 38px;
  font-size: 16px;
  line-height: 28px;
  border-radius: 24px;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e9eef2;
  border-color: #e9eef2;
  color: #ffffff;
}

.c0:hover:not(:disabled),
.c0:focus:not(:disabled) {
  background-color: #67cff3;
  border-color: #67cff3;
}

.c0:active:not(:disabled) {
  background-color: #147ec1;
  border-color: #147ec1;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  display="inline-block"
>
  Hello
</button>
`;
